File: python3/kth_largest_element_in_an_array.py
"""
Space   : O(1)
Time    : O(n log n)
"""


class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        return sorted(nums)[-k]

    # O(N) Time, worth to consider
    # but still slower than sorted
    def findKthLargestPartition(self, nums: List[int], k: int) -> int:
        return self.findKthSmallest(nums, len(nums)+1-k)

    def findKthSmallest(self, nums, k):
        if nums:
            pos = self.partition(nums, 0, len(nums)-1)
            if k > pos+1:
                return self.findKthSmallest(nums[pos+1:], k-pos-1)
            elif k < pos+1:
                return self.findKthSmallest(nums[:pos], k)
            else:
                return nums[pos]

    def partition(self, nums, l, r):
        low = l
        while l < r:
            if nums[l] < nums[r]:
                nums[l], nums[low] = nums[low], nums[l]
                low += 1
            l += 1
        nums[low], nums[r] = nums[r], nums[low]
        return low

File: golang/kth_largest_element_in_an_array.go
package main

/*
Space	: O(n)
Time	: O(n)
*/

func findKthLargest(nums []int, k int) int {
	return findKthSmallest(nums, len(nums)+1-k)
}

func findKthSmallest(nums []int, k int) int {
	pos := partitionArray(nums, 0, len(nums)-1)
	if k > pos+1 {
		return findKthSmallest(nums[pos+1:], k-pos-1)
	} else if k < pos+1 {
		return findKthSmallest(nums[:pos], k)
	}
	return nums[pos]
}

func partitionArray(nums []int, l int, r int) int {
	low := l
	for l < r {
		if nums[l] < nums[r] {
			nums[l], nums[low] = nums[low], nums[l]
			low++
		}
		l++
	}
	nums[low], nums[r] = nums[r], nums[low]
	return low
}
